### PR TITLE
Cellular Automaton for quadruplet pixel seeding

### PR DESCRIPTION
**Pull request by felicepantaleo**
Phase 1 quadruplet seeding using Cellular automaton, disabled by default.
You can find the talk here:
https://indico.cern.ch/event/542270/
you can customise the process by adding:

`# customisation of the process.

from RecoTracker.Configuration.customiseForQuadrupletsByCellularAutomaton import customiseForQuadrupletsByCellularAutomaton

process = customiseForQuadrupletsByCellularAutomaton(process)
`

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    
+    
+private:
+    edm::EDGetTokenT<SeedingLayerSetsHits> theSeedingLayerToken;
+
+    LayerCacheType theLayerCache;
+
+    std::unique_ptr<SeedComparitor> theComparitor;
+
+    class QuantityDependsPtEval {
```
Comment (makortel) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.h]: Maybe we should abstract `QuantityDependsPtEval` out of being inner class of `CAHitQuadrupletGenerator` and `PixelQuadrupletGenerator` instead of copy-paste (for me it's fine to cleanup in a later PR).

File: RecoTracker/TkHitPairs/src/HitPairGeneratorFromLayerPair.cc [hunk 2]
```diff
@@ -109,6 +109,29 @@ HitDoublets HitPairGeneratorFromLayerPair::doublets( const TrackingRegion& regio
 
 }
 
+
+HitDoublets HitPairGeneratorFromLayerPair::doublets( const TrackingRegion& region,
+						    const edm::Event & iEvent, const edm::EventSetup& iSetup, const Layer& innerLayer, const Layer& outerLayer) {
```
Comment (makortel) [inline RecoTracker/TkHitPairs/src/HitPairGeneratorFromLayerPair.cc]: It would be possible to reduce copy-paste by making the overload above to call this overload (but again I'm fine with a later cleanup PR).

**Comment by makortel** [comment]
Thanks @felicepantaleo for the update, the overall structure looks good to me now.

**Comment by slava77** [comment]
@felicepantaleo 
please let me know if you can follow up to comments from @makortel in this PR.
Thank you.

**Comment by felicepantaleo** [comment]
Thanks @makortel @slava77 , I will clean up duplicate code in a later PR next week.

**Comment by VinInn** [comment]
@cmsbuild , please test

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h [hunk 3]
```diff
@@ -0,0 +1,233 @@
+#ifndef CACELL_H_
+#define CACELL_H_
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: please match the include guard to the package/file name 
```RecoPixelVertexing_PixelTriplets_CACell_h ``` (CACELL is too short)

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    using Hit = RecHitsSortedInPhi::Hit;
+    using CAntuplet = std::vector<CACell*>;
+
+    CACell(const HitDoublets* doublets, int doubletId, const unsigned int cellId, const int innerHitId, const int outerHitId) :
+    theCAState(0), theInnerHitId(innerHitId), theOuterHitId(outerHitId), theCellId(cellId), hasSameStateNeighbors(0), theDoublets(doublets), theDoubletId(doubletId),
+    theInnerR(doublets->r(doubletId, HitDoublets::inner)), theOuterR(doublets->r(doubletId, HitDoublets::outer)),
+    theInnerZ(doublets->z(doubletId, HitDoublets::inner)), theOuterZ(doublets->z(doubletId, HitDoublets::outer)) {
+    }
+
+    unsigned int get_cell_id() const {
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: bouncing caps is a more typical CMSSW method naming style.
underscore-separated in ~~ OK.
In case you get to refactoring.

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        }
+    }
+
+    bool are_aligned_RZ(const CACell* otherCell, const float ptmin, const float thetaCut) const {
+
+        float r1 = otherCell->get_inner_r();
+        float z1 = otherCell->get_inner_z();
+        float distance_13_squared = (r1 - theOuterR)*(r1 - theOuterR) + (z1 - theOuterZ)*(z1 - theOuterZ);
+        float tan_12_13_half = fabs(z1 * (theInnerR - theOuterR) + theInnerZ * (theOuterR - r1) + theOuterZ * (r1 - theInnerR)) / distance_13_squared;
+        return tan_12_13_half * ptmin <= thetaCut;
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: division (and a potential FPE) could be avoided by ``` return a*b < c*d ``` instead of ``` a/c*b < d ```

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        auto x2 = get_inner_x();
+        auto y2 = get_inner_y();
+
+        auto x3 = get_outer_x();
+        auto y3 = get_outer_y();
+
+        auto precision = 0.5f;
+        auto offset = x2 * x2 + y2*y2;
+
+        auto bc = (x1 * x1 + y1 * y1 - offset) / 2.f;
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: *0.5f instead ?

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+        auto idet = 1.f / det;
+
+        auto x_center = (bc * (y2 - y3) - cd * (y1 - y2)) * idet;
+        auto y_center = (cd * (x1 - x2) - bc * (x2 - x3)) * idet;
+
+        auto radius = std::sqrt((x2 - x_center)*(x2 - x_center) + (y2 - y_center)*(y2 - y_center));
+        auto centers_distance_squared = (x_center - region_origin_x)*(x_center - region_origin_x) + (y_center - region_origin_y)*(y_center - region_origin_y);
+
+        auto minimumOfIntesectionRange = (radius - region_origin_radius)*(radius - region_origin_radius) - phiCut;
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: was it "inte_r_section" or intesection?

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+  ca.find_ntuplets(foundQuadruplets, 4);
+
+  const QuantityDependsPtEval maxChi2Eval = maxChi2.evaluator(es);
+
+
+  // re-used thoughout, need to be vectors because of RZLine interface
+  std::vector<float> bc_r(4), bc_z(4), bc_errZ(4);
+
+  declareDynArray(GlobalPoint, 4, gps);
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: why dynArray if the size (4) is known at compile time?
use just std::array
Same for the vectors bc_r etc

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+  // Loop over quadruplets
+  for (unsigned int quadId = 0; quadId < numberOfFoundQuadruplets; ++quadId)
+  {
+
+    auto isBarrel = [](const unsigned id) -> bool
+    {
+      return id == PixelSubdetector::PixelBarrel;
+    };
+
+    gps[0] = foundQuadruplets[quadId][0]->get_inner_hit()->globalPosition();
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: multiple copy-paste of ``` foundQuadruplets[quadId][0]->get_inner_hit() ``` can be reduced by using an ``` auto const& ahit  = foundQuadruplets[quadId][0]->get_inner_hit();```

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+      }
+    }
+
+    float chi2 = std::numeric_limits<float>::quiet_NaN();
+    // TODO: Do we have any use case to not use bending correction?
+    if (useBendingCorrection)
+    {
+      // Following PixelFitterByConformalMappingAndLine
+      const float simpleCot = ( gps.back().z() - gps.front().z() ) / (gps.back().perp() - gps.front().perp() );
+      const float pt = 1 / PixelRecoUtilities::inversePt(abscurv, es);
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: ```1.0f``` to stay in floats

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+      for (int i=0; i < 4; ++i)
+      {
+        const GlobalPoint & point = gps[i];
+        const GlobalError & error = ges[i];
+        bc_r[i] = sqrt( sqr(point.x() - region.origin().x()) + sqr(point.y() - region.origin().y()) );
+        bc_r[i] += pixelrecoutilities::LongitudinalBendingCorrection(pt, es)(bc_r[i]);
+        bc_z[i] = point.z() - region.origin().z();
+        bc_errZ[i] =  (barrels[i]) ? sqrt(error.czz()) : sqrt( error.rerr(point) ) * simpleCot;
+      }
+      RZLine rzLine(bc_r, bc_z, bc_errZ);
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: funny: the only reason to call sqrt in bc_errZ is to pass it to the constructor which immediately takes a square of this.

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.h [hunk 12]
```diff
@@ -0,0 +1,131 @@
+#ifndef CAHitQuadrupletGenerator_H
+#define CAHitQuadrupletGenerator_H
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.h]: use a more specific include guard, please.

File: RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+  auto outerLayerId = innerLayerId + 1;
+  auto numberOfDoublets = doublets[layerPairId]->size ();
+
+  isOuterHitOfCell[outerLayerId].resize(fourLayers[outerLayerId].hits().size());
+  theFoundCellsPerLayer[layerPairId].reserve (numberOfDoublets);
+  for (unsigned int i = 0; i < numberOfDoublets; ++i)
+  {
+    theFoundCellsPerLayer[layerPairId].emplace_back (doublets[layerPairId], i,  cellId,  doublets[layerPairId]->innerHitId(i), doublets[layerPairId]->outerHitId(i));
+
+    isOuterHitOfCell[outerLayerId][doublets[layerPairId]->outerHitId(i)].push_back (&(theFoundCellsPerLayer[layerPairId][i]));
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc]: better use refs or temporaries to avoid repeated  [layerPairId] position computation (unclear if it gets optimized out)

Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc]: is there a .reserve for these push_backs?

File: RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+  for (CACell* root_cell : theRootCells)
+  {
+    tmpNtuplet.clear();
+    tmpNtuplet.push_back(root_cell);
+    root_cell->find_ntuplets (foundNtuplets, tmpNtuplet, minHitsPerNtuplet);
+  }
+
+}
+
+template class CellularAutomaton<4>;
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc]: add a new line, please

File: RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.h [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+private:
+
+
+
+    //for each hit in each layer, store the pointers of the Cells of which it is outerHit
+    std::array<std::vector<std::vector<CACell*> >, theNumberOfLayers> isOuterHitOfCell;
+    std::array<std::vector<CACell>, theNumberOfLayers> theFoundCellsPerLayer;
+
+    std::vector<CACell*> theRootCells;
+    std::vector<std::vector<CACell*> > theNtuplets;
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.h]: how much does this grow/vary in a typical execution?
It may be better to linearize into 1D vector or have a fixed cap size (array) on the outer index.

File: RecoPixelVertexing/PixelTriplets/plugins/SealModule.cc [hunk 16]
```diff
@@ -25,3 +25,6 @@ DEFINE_EDM_PLUGIN(OrderedHitsGeneratorFactory, CombinedHitQuadrupletGenerator, "
 #include "RecoPixelVertexing/PixelTriplets/interface/HitQuadrupletGeneratorFromTripletAndLayersFactory.h"
 #include "PixelQuadrupletGenerator.h"
 DEFINE_EDM_PLUGIN(HitQuadrupletGeneratorFromTripletAndLayersFactory, PixelQuadrupletGenerator, "PixelQuadrupletGenerator");
+
+#include "CAHitQuadrupletGenerator.h"
+DEFINE_EDM_PLUGIN(OrderedHitsGeneratorFactory, CAHitQuadrupletGenerator, "CAHitQuadrupletGenerator");
```
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/SealModule.cc]: please add a new line.

Technically, the DEFINE_EDM_PLUGIN line can be added in CAHitQuadrupletGenerator.h
SealModule.cc file is not required

**Comment by slava77** [comment]
@felicepantaleo please edit the PR title to reflect that this code applies to pixel seeding

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc (hunk 8 above)
Comment (makortel) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: Because of interfaces of `RZLine` and `FastCircleFit`, I presume (ok, nowadays they have also the pointer-interface, but I'd leave them to be improved later). Note that the `PixelQuadrupletGenerator` uses exactly the same types.

File: RecoPixelVertexing/PixelTriplets/plugins/SealModule.cc (hunk 16 above)
Comment (makortel) [inline RecoPixelVertexing/PixelTriplets/plugins/SealModule.cc]: I'd leave `DEFINE_EDM_PLUGIN` to the `SealModule.cc` for consistency because all other plugins are defined there (if they should be moved elsewhere, better done in a separate PR).

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc (hunk 8 above)
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: ok for bc_, which have to go through the vector interface.
gps, ges, barrels are passed as raw pointers. This can be done with array::data()

File: RecoPixelVertexing/PixelTriplets/plugins/SealModule.cc (hunk 16 above)
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/SealModule.cc]: OK

**Comment by slava77** [comment]
@felicepantaleo please follow up on the comments from the code review

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h (hunk 3 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: done

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h (hunk 4 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: done

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h (hunk 5 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: done

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h (hunk 6 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: done

File: RecoPixelVertexing/PixelTriplets/plugins/CACell.h (hunk 7 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CACell.h]: done ;)

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc (hunk 8 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: I'm keeping this unmodified. It will be object of a future and more extended PR.

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc (hunk 9 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: done.

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc (hunk 10 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: done.

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc (hunk 11 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc]: yep. To be changed in a later PR.

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.h (hunk 12 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.h]: done

File: RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc (hunk 13 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc]: There is no reserve because we haven't still measured how many doublets start from the same hit wrt PU 35->70.
Do you suggest to reserve a reasonable size and then modify it later, or to leave like it is and then make a pull request with the final size?

File: RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc (hunk 14 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc]: done

File: RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.h (hunk 15 above)
Comment (felicepantaleo) [inline RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.h]: Like in a previous comment: there is no typical execution yet. we will study what's a good size and we will have a fixed-size array.

**Comment by VinInn** [comment]
@cmsbuild , please test

File: RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc (hunk 13 above)
Comment (slava77) [inline RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc]: I was expecting something reasonable to not start off from "1".
There is also a dynamic machinery from Vincenzo (FWCore/Utilities/interface/RunningAverage.h )
I suppose you have tested this in PU35.
So, you know how far this goes.
If it stays at O(1), doesn't matter.
If it's O(100), it's likely ~5-8 reallocations while growing the vector.

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by slava77** [comment]
@cmsbuild please test
maybe no DAS errors this time

**Comment by slava77** [comment]
@smuzaffar @iahmad-khan 
the 1000.0 and 1001.0 fail in the baseline as well 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_8_1_X_2016-07-12-2300/slc6_amd64_gcc530/-GenuineIntel/matrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/
I recall there was a check that would skip failures like this and let the comparisons complete.
Why is it not picked up here?

**Comment by smuzaffar** [comment]
@slava77 , that check is applied when we run comparison.

**Comment by slava77** [comment]
On 7/13/16 1:05 AM, Malik Shahzad Muzaffar wrote:
> @slava77 <https://github.com/slava77> , that check is applied when we
> run comparison.
> 

I see.
So, this logic of failed test in baseline worked only if all tests
passed in the current PR test.
If a workflow fails in a PR test, the comparisons are not run at all.

https://github.com/cms-sw/cms-bot/blob/76b032bc3dedb89a5e95cb5e57a2e4b863329f52/run-pr-tests#L490
COMPARISON;NOTRUN is set if there is any ERROR in this PR log file.
We should ignore a failure in the PR test if it also fails in the IB
in the decision to run comparisons.

What do you think?

**Comment by slava77** [comment]
@cmsbuild please test
I see that some PRs came back with tests passing

**Comment by slava77** [comment]
- local tests of 10224.0 (TTbar PU in 2017 setup) with CA enabled shows changes roughly in line with the presentation in the RECO meeting: some loss of tracks at low pt and a small increase in fakes.

Tracks not reconstructed with quad step are picked up with "later" (lowPtTripletStep and detachedTripletStep) iterations
all_sign738-histatvsorig-histat_ttbar13tevpu2017wf10224p0c_recotracks_generaltracks__reco_obj_algo

Somewhat minor decrease in generalTracks yield 
all_sign738-histatvsorig-histat_ttbar13tevpu2017wf10224p0c_log10recotracks_generaltracks__reco_obj_pt
(via algo?) leads to a more significant effect in PF charged hadrons
all_sign738-histatvsorig-histat_ttbar13tevpu2017wf10224p0c_log10recopfcandidates_particleflow__reco_obj_pt15
This all is at low enough pt that downstream quantities are not really significantly affected

Timing changes are most significant only in iter tracking steps and roughly reflects an effect of more or less tracks reconstructed with different iterations
```
deltaSelf       deltaJob          baseline                    new
   -0.645081      -0.22%       471.36 ms/ev ->       241.45 ms/ev lowPtQuadStepSeeds
   -0.043828      -0.03%       624.76 ms/ev ->       597.96 ms/ev lowPtQuadStepTrackCandidates
   -0.183505      -0.06%       401.90 ms/ev ->       334.35 ms/ev lowPtQuadStepTracks

   -0.389778      -0.06%       186.71 ms/ev ->       125.80 ms/ev initialStepSeedsPreSplitting
   +0.230274      +0.19%       763.79 ms/ev ->       962.56 ms/ev initialStepTrackCandidatesPreSplitting
   +0.034882      +0.02%       462.97 ms/ev ->       479.41 ms/ev initialStepTracksPreSplitting

   -0.384804      -0.16%       511.17 ms/ev ->       346.21 ms/ev detachedQuadStepSeeds
   +0.291762      +0.03%        95.73 ms/ev ->       128.43 ms/ev detachedQuadStepTrackCandidates

   +0.051849      +0.02%       406.28 ms/ev ->       427.90 ms/ev lowPtTripletStepSeeds
   +0.120556      +0.08%       643.31 ms/ev ->       725.84 ms/ev lowPtTripletStepTrackCandidates
   +0.063103      +0.01%       229.32 ms/ev ->       244.26 ms/ev lowPtTripletStepTracks

   +0.052302      +0.01%       198.67 ms/ev ->       209.34 ms/ev detachedTripletStepSeeds
   +0.090310      +0.03%       304.16 ms/ev ->       332.93 ms/ev detachedTripletStepTrackCandidates
   +0.058050      +0.01%       176.18 ms/ev ->       186.71 ms/ev detachedTripletStepTracks

   +0.028554      +0.02%       639.10 ms/ev ->       657.62 ms/ev pixelLessStepSeeds

```

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/15102 	b8dca54
- new code with CA implementation, in line with the description and a presentation in the RECO meeting today. The new algo is disabled by default, but can be enabled with ``` --customise RecoTracker/Configuration/customiseForQuadrupletsByCellularAutomaton.customiseForQuadrupletsByCellularAutomaton``` passed to the cmsDriver.py or runTheMatrix.py
- jenkins tests pass up to compilation and matrix tests running (failures related to DAS blocked the comparisons)
- local tests confirm no changes in the standard workflows
- tests with the CA enabled show roughly expected behavior, without a major effect on objects downstream of tracking
